src/facade/dragonfly_connection.cc

```diff
@@ -934,34 +953,51 @@ void Connection::ConnectionFlow(FiberSocketBase* peer) {
   }
 }
 
-void Connection::DispatchCommand(bool has_more, absl::FunctionRef<void()> dispatch_sync,
-                                 absl::FunctionRef<MessageHandle()> dispatch_async) {
+void Connection::DispatchSingle(bool has_more, absl::FunctionRef<void()> invoke_cb,
+                                absl::FunctionRef<MessageHandle()> cmd_msg_cb) {
   // Avoid sync dispatch if we can interleave with an ongoing async dispatch
   bool can_dispatch_sync = !cc_->async_dispatch;
 
   // Avoid sync dispatch if we already have pending async messages or
   // can potentially receive some (subscriptions > 0)
-  if (dispatch_q_.size() > 0 || cc_->subscriptions > 0)
-    can_dispatch_sync = false;
+  if (can_dispatch_sync && (!dispatch_q_.empty() || cc_->subscriptions > 0)) {
+    DCHECK(queue_backpressure_ == &tl_queue_backpressure_);
+    if (stats_->dispatch_queue_bytes < queue_backpressure_->pipeline_buffer_limit) {
+      can_dispatch_sync = false;
+    } else {
+      fb2::NoOpLock noop;
+      queue_backpressure_->pipeline_cnd.wait(noop, [this] {
+        bool res = stats_->dispatch_queue_bytes < queue_backpressure_->pipeline_buffer_limit ||
+                   (dispatch_q_.empty() && !cc_->async_dispatch) || cc_->conn_closing;
+        return res;
+      });
+      if (cc_->conn_closing)
+        return;
+
+      // Prefer sending synchronous request if possible (can_dispatch_sync=false),
+      // to reduce the memory pressure.
+      has_more = false;
+      if (cc_->async_dispatch || !dispatch_q_.empty() || cc_->subscriptions > 0) {
+        can_dispatch_sync = false;
+      }
+    }
+  }
```

Comment:
I understand the idea, but the code is somewhat confusing. You re-use conditions like `can_dispatch_sync` for control, same with `has_more` - that's a propery of the current connection state 🙂 And after waking up you re-evaluate the conditoons from above 🤯 

Maybe we could first make conclusions based on connection state, then with new variables determine how we dispatch and after waking, we can actually call the function recursively again

Reply:
I am not a fan of recursive calls. I am open for other suggestions on how to improve the code.

Reply:
Recursive calls was my last suggestion. The main problem is confusing variable names. Again - has_more is a connection property - not a good variable name to control behaviour or reassign